src/ppom-controller.ts

```diff
@@ -0,0 +1,387 @@
+      const fileData = await this.#fetchBlob(fileUrl);
+      if (!fileData) {
+        throw new Error('Failed to fetch file data');
+      }
+
+      newFiles.push({
+        data: fileData,
+        ...fileVersionInfo,
+      });
+    }
+
+    return newFiles;
+  }
+
+  /*
+   * Fetch the version info from the PPOM cdn.
+   *  update the version info in state.
+   */
+  async #updateVersionInfo() {
+    const versionInfo = await this.#fetchVersionInfo(PPOM_VERSION_PATH);
+    if (!versionInfo) {
+      throw new Error('Failed to fetch version info');
+    }
+
+    this.update({
+      versionInfo,
+      lastFetched: Date.now(),
+    });
+  }
+
+  /**
+   * Conditionally update the ppom configuration.
+   *
+   * If the ppom configuration is out of date, this function will call `updatePPOM`
+   * to update the configuration.
+   */
+  async #maybeUpdatePPOM() {
+    if (await this.#shouldUpdate()) {
+      await this.updatePPOM();
+    }
+
+    this.update(this.state);
+  }
+
+  /*
+   * Fetch the blob from the PPOM cdn.
+   */
+  async #fetchBlob(input: string): Promise<ArrayBuffer | null> {
+    const response = await safelyExecute(
+      async () => fetch(input, { cache: 'no-cache' }),
+      true,
+    );
+
+    switch (response?.status) {
+      case 200: {
+        return await response.arrayBuffer();
+      }
+
+      default: {
+        return null;
+      }
+    }
+  }
+
+  /*
+   * Fetch the version info from the PPOM cdn.
+   */
+  async #fetchVersionInfo(input: string): Promise<PPOMVersionResponse | null> {
+    const response = await safelyExecute(
+      async () => fetch(input, { cache: 'no-cache' }),
+      true,
+    );
+    switch (response?.status) {
+      case 200: {
+        return response.json();
+      }
+
+      default: {
+        return null;
+      }
+    }
+  }
+
+  /*
+   * Send a JSON RPC request to the provider.
+   * This method is used by the PPOM to make requests to the provider.
+   */
+  async #jsonRpcRequest(req: any): Promise<any> {
+    return new Promise((resolve) => {
+      this.#provider.sendAsync(req, (_err: any, res: any) => {
+        resolve(res);
+      });
+    });
+  }
+
+  /*
+   * Initialize the PPOM.
+   * This function will be called when the PPOM is first used.
+   * or when the PPOM is out of date.
+   * It will load the PPOM data from storage and initialize the PPOM.
+   */
+  async #getPPOM(): Promise<PPOM> {
+    await ppomInit(this.#storage.readFile(PPOM_BLOB_NAME, ''));
+    const data = await this.#storage.readFile(
+      PPOM_DATA_NAME,
+      this.state.newChainId,
+    );
+
+    return new PPOM(this.#jsonRpcRequest.bind(this), new Uint8Array(data));
+  }
+
+  /**
+   * Use the PPOM.
+   * This function receives a callback that will be called with the PPOM.
+   * The callback will be called with the PPOM after it has been initialized.
+   *
+   * @param callback - Callback to be invoked with PPOM.
+   */
+  async use(callback: (ppom: PPOM) => Promise<any>): Promise<any> {
```

Comment:
Will the return type will always match the callback?

If so, should we use a simple template to make that explicit in the typing such as:

`async use<T>(callback: (ppom: PPOM) => Promise<T>): Promise<T>`

Reply:
Type of the 2 promises returned above are different.